src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/LocalCIDockerService.java

```diff
@@ -282,15 +291,102 @@ private void checkImageArchitecture(String imageName, InspectImageResponse inspe
+     *           - We need to iterate over the map entries since don't remove the oldest image from the map.
+     */
+
+    @Scheduled(fixedRateString = "${artemis.continuous-integration.image-cleanup.disk-space-check-interval-ms:3600000}", initialDelayString = "${artemis.continuous-integration.image-cleanup.disk-space-check-interval-ms:3600000}")
+    public void checkUsableDiskSpaceThenCleanUp() {
+        if (!imageCleanupEnabled) {
+            return;
+        }
+        try {
+            // Get the Docker root directory to check disk space.
+            File dockerRootDirectory = new File(Objects.requireNonNullElse(dockerClient.infoCmd().exec().getDockerRootDir(), "/"));
+            long usableSpace = dockerRootDirectory.getUsableSpace();
+
+            long threshold = convertMegabytesToBytes(imageCleanupDiskSpaceThresholdMb);
+
+            if (usableSpace >= threshold) {
+                return;
+            }
+
+            // Get map of docker images and their last build dates
+            IMap<String, ZonedDateTime> dockerImageCleanupInfo = hazelcastInstance.getMap("dockerImageCleanupInfo");
+
+            // Get unused images
+            Set<String> unusedImages = getUnusedDockerImages();
+
+            // Get a sorted list of images by last build date
+            // We cast to ArrayList since we need the list to be mutable
+            List<Map.Entry<String, ZonedDateTime>> sortedImagesByLastBuildDate = dockerImageCleanupInfo.entrySet().stream().sorted(Map.Entry.comparingByValue()).toList();
+            List<Map.Entry<String, ZonedDateTime>> mutableSortedImagesByLastBuildDate = new java.util.ArrayList<>(sortedImagesByLastBuildDate);
+
+            if (mutableSortedImagesByLastBuildDate.isEmpty()) {
+                return;
+            }
+
+            int deleteAttempts = 5;
+            int totalAttempts = mutableSortedImagesByLastBuildDate.size(); // We limit the total number of attempts to avoid infinite loops
+            Map.Entry<String, ZonedDateTime> oldestImage = mutableSortedImagesByLastBuildDate.getFirst();
+            while (oldestImage != null && usableSpace < threshold && deleteAttempts > 0 && totalAttempts > 0) {
+                if (unusedImages.contains(oldestImage.getKey())) {
+                    log.info("Deleting docker image {}", oldestImage.getKey());
+                    try {
+                        dockerClient.removeImageCmd(oldestImage.getKey()).exec();
+                        usableSpace = dockerRootDirectory.getUsableSpace();
+                        deleteAttempts--;
+                    }
+                    catch (NotFoundException e) {
+                        log.warn("Docker image {} not found during cleanup", oldestImage.getKey());
+                    }
+                }
+                mutableSortedImagesByLastBuildDate.remove(oldestImage);
+                oldestImage = mutableSortedImagesByLastBuildDate.getFirst();
+                totalAttempts--;
```

Comment:
Something that I just noticed: I think it is more common / intuitive to increment than decrement within loops.
Setting a constant max value in the while loop for  `deleteAttempts` and for `totalAttempts`. For me the while loop condition would be easier to read then.

Since I already had some comments in the code and it didn't bother me on the first review, I do not think that this is something that needs to be addressed here but is rather something for future implementations :)